### PR TITLE
Add option to list the emitted files as part of the compiler output.

### LINKED ISSUE #3650: Possibly add a new compiler option to list output files.
I'm adding support for TypeScript in the [Qbs](http://doc.qt.io/qbs) build system. Due to the way we track dependencies, we require knowing the list of output files a particular tool will produce when invoked with a particular set of arguments.

Currently, this can be difficult to determine since it requires actually parsing the TypeScript source code (i.e. external modules) rather than simply looking at filenames (as in the case of a C compiler, for example).

I'd like to request the addition of an additional option, `--exportDependencyInfo`, which will output in JSON format a list of all files that will be written by the compiler for a particular command line invocation. This would be used in combination with --noEmit, for example:

    tsc --noEmit -exportDependencyInfo --declaration --sourceMap --module amd --outDir /var/foo/bar \
        /var/public/a.ts \
        /var/public/b.ts

Expected output (assume b.ts imports an external module c):

    [
        {"filePath": "/var/foo/bar/a.js"},
        {"filePath": "/var/foo/bar/a.js.map"},
        {"filePath": "/var/foo/bar/a.d.ts"},
        {"filePath": "/var/foo/bar/b.js"},
        {"filePath": "/var/foo/bar/b.js.map"},
        {"filePath": "/var/foo/bar/b.d.ts"},
        {"filePath": "/var/foo/bar/c.js"},
        {"filePath": "/var/foo/bar/c.js.map"},
        {"filePath": "/var/foo/bar/c.d.ts"},
    ]

I notice there's a --listFiles option documented here: https://github.com/Microsoft/TypeScript/wiki/Compiler%20Options Will this become a documented option in a future version? If so, this could be a possible solution to our problem, as we could find the common base path of the input files printed by --listFiles, and calculate the corresponding output filenames.

However, it would be ideal if the compiler itself did this, to avoid any potential discrepancy.

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Fixes #3650

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -1015,7 +1015,7 @@ namespace ts {
             let declarationDiagnostics: Diagnostic[] = [];
 
             if (options.noEmit) {
-                return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emitSkipped: true };
+                return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emittedFiles: [], emitSkipped: true };
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Consider `undefined` here

File: src/compiler/program.ts [hunk 2]
```diff
@@ -1015,7 +1015,7 @@ namespace ts {
             let declarationDiagnostics: Diagnostic[] = [];
 
             if (options.noEmit) {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Why are there two places where we check `noEmit`? One in `program.ts` and one in `emitter.ts`?

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: @paulvanbrenk says it could be the language service (i.e. Compile-on-Save)

File: src/compiler/tsc.ts [hunk 3]
```diff
@@ -14,6 +14,17 @@ namespace ts {
         }
     }
 
+    function reportEmittedFiles(files: string[], host: CompilerHost): void {
+        if (!files || files.length == 0) {
+            return;
+        }
+
+        for (const file of files) {
+            const message = `TSFILE: ${file}${sys.newLine}`;
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Don't really need a temp var

**Comment by DanielRosenwasser** [comment]
Looks good to me, just have @vladima and/or @sheetalkamat check.

File: src/compiler/tsc.ts [hunk 4]
```diff
@@ -14,6 +14,18 @@ namespace ts {
         }
     }
 
+    function reportEmittedFiles(files: string[], host: CompilerHost): void {
+        if (!files || files.length == 0) {
+            return;
+        }
+
+        for (const file of files) {
+            const filepath = getNormalizedAbsolutePath(file, sys.getCurrentDirectory());
```
Comment (vladima) [inline src/compiler/tsc.ts]: create a local for the call `sys.getCurrentDirectory()`

**Comment by vladima** [comment]
:+1: with one comment

File: src/compiler/commandLineParser.ts [hunk 5]
```diff
@@ -332,6 +332,11 @@ namespace ts {
             description: Diagnostics.Do_not_emit_use_strict_directives_in_module_output
         },
         {
+            name: "listEmittedFiles",
+            type: "boolean",
+            description: Diagnostics.List_emitted_files
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: do not add the description to avoid having it in the help. most users do not care about this flag

**Comment by mhegazy** [comment]
:+1:

**Comment by jakepetroules** [comment]
This is awesome! I've been wanting this for a while; it will allow me to remove a workaround in the [Qbs build system](http://qbs.io) that uses the TypeScript compiler APIs to achieve this.

However, I'm a little concerned about the stability of the output (given that it is unstructured plain text) and it being mixed into standard out with anything else that the compiler might decide to write there in the future.

Would it be possible to add an optional argument to --listEmittedFiles which allows a file path to be specified, where the TypeScript compiler will write the generated filenames as a JSON array? Especially if this is intended to be machine readable (and it is), we should have stability guarantees for the output.

**Comment by paulvanbrenk** [comment]
@jakepetroules something like this: #7975

**Pull request merged**